taskcluster/ac_taskgraph/loader/build_config.py

```diff
@@ -13,7 +17,113 @@
+# maintains the latter mapping.
+FILE_PATTERNS_TO_AFFECTED_COMPONENTS = {
+    # Just a placeholder example
+    #"^.*changelog.md.*$": ["build-browser-awesomebar"],
+    "^build.gradle$": ALL_COMPONENTS,
+}
+
+def get_files_changed_pr(base_repository, pull_request_number):
+    url = base_repository.replace("github.com", "api.github.com/repos")
+    url += "/pulls/%s/files" % pull_request_number
+
+    r = requests.get(url, timeout=60)
+    r.raise_for_status()
+    files = [f["filename"] for f in r.json()]
+    return files
+
+def get_files_changed_push(base_repository, base_rev, head_rev):
+    url = base_repository.replace("github.com", "api.github.com/repos")
+    url += "/compare/"
+    url += "%s...%s" % (base_rev, head_rev)
+
+    r = requests.get(url, timeout=60)
+    r.raise_for_status()
+    files = [f["filename"] for f in r.json().get("files")]
+    return files
+
+def get_components_changed(files_changed):
+    """Translate a list of files changed into a list of components. Eg:
+        [
+            "components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt",
+            "components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegateTest.kt",
+        ]
+        ->
+        {service-sync-logins}
+    """
+    return set(["-".join(f.split("/")[1:3]) for f in files_changed if f.startswith("components")])
+
+
+cached_deps = {}
+
+def get_deps_for_component(component):
+    """Return the full list of local upstream dependencies of a component."""
+    # Gradle is expensive to run. Cache the results for each component.
+    # TODO: replace me with a functools.lru_cache decorator when this runs on py3
+    if component in cached_deps:
+        return cached_deps[component]
+
+    deps = set()
+    # Parsing output like this is not ideal but bhearsum couldn't find a way
+    # to get the dependencies printed in a better format. If we could convince
+    # gradle to spit out JSON that would be much better.
+    for line in subprocess.check_output(['./gradlew', '%s:dependencies' % component, '--configuration', 'implementation']).splitlines():
+        if line.startswith("+--- project"):
+            deps.add(line.split(" ")[2])
+
+    cached_deps[component] = deps
+    # So far, we only have the direct local upstream dependencies. We need to look at
+    # each of those to find _their_ upstreams as well.
+    for d in deps.copy():
+        deps.update(get_deps_for_component(d))
+
+    return deps
+
+def get_affected_components(files_changed):
+    affected_components = set()
+
+    # First, find the list of changed components as well as all of their
+    # local upstreams.
+    for c in get_components_changed(files_changed):
+        affected_components.add(c)
+        affected_components.update(get_deps_for_component(c))
+
+    # Then, look for any other affected components based on the files changed.
```

Comment:
Good thinking! On tope of my head, I don't know if we should track any other files. Any idea @pocmo?

Reply:
I'm hoping this list will be short (and I'd be happy for it to be empty), but I suspect at least some top level gradle files will need be included...

Reply:
Of the top of my head I'd say a full rebuild should be needed for:
* Changes in `buildSrc`
* Changes in root `build.gradle` and `settings.gradle`
* Changes in gradle.properties
* Changes in gradle/wrapper/gradle-wrapper.properties

Reply:
> Just to clarify and make sure I understand it correctly...
> 
> This patch will:
> 
>     * Schedule only the tasks for components that should be affected by a push

Correct.
> 
> This patch will NOT:
> 
>     * Share build artifacts between tasks (when building component C depending on A and B then the task for C will also build A and B).

Correct - the internals of the build tasks and their task dependencies are totally unchanged. This patch _only_ prunes out tasks from the existing list of tasks.

>     * Rebuild the dependency tree in taskcluster, e.g. if A depends on B and B depends on C then we won't build C first, then B then A, but instead build A, B, C in parallel (which makes sense without sharing build artifacts)

Right. We're extracting the dependency tree from gradle in the decision task. I have no desire to maintain a parallel dependency tree :).

I'd love to see the latter two things happen at some point, but I think it's much more work for much less of a win - so I have no intention to look at them in the near future.